src/cmd/services/m3coordinator/ingest/carbon/ingest.go

```diff
@@ -0,0 +1,265 @@
+)
+
+// Options configures the ingester.
+type Options struct {
+	InstrumentOptions instrument.Options
+	WorkerPool        xsync.PooledWorkerPool
+	Timeout           time.Duration
+}
+
+// Validate validates the options struct.
+func (o *Options) Validate() error {
+	if o.InstrumentOptions == nil {
+		return errIOptsMustBeSet
+	}
+
+	if o.WorkerPool == nil {
+		return errWorkerPoolMustBeSet
+	}
+
+	return nil
+}
+
+// NewIngester returns an ingester for carbon metrics.
+func NewIngester(
+	downsamplerAndWriter ingest.DownsamplerAndWriter,
+	opts Options,
+) (m3xserver.Handler, error) {
+	err := opts.Validate()
+	if err != nil {
+		return nil, err
+	}
+
+	return &ingester{
+		downsamplerAndWriter: downsamplerAndWriter,
+		opts:                 opts,
+		logger:               opts.InstrumentOptions.Logger(),
+		metrics: newCarbonIngesterMetrics(
+			opts.InstrumentOptions.MetricsScope()),
+	}, nil
+}
+
+type ingester struct {
+	downsamplerAndWriter ingest.DownsamplerAndWriter
+	opts                 Options
+	logger               log.Logger
+	metrics              carbonIngesterMetrics
+}
+
+func (i *ingester) Handle(conn net.Conn) {
+	var (
+		wg     = sync.WaitGroup{}
+		s      = carbon.NewScanner(conn, i.opts.InstrumentOptions)
+		logger = i.opts.InstrumentOptions.Logger()
+	)
+
+	logger.Debug("handling new carbon ingestion connection")
+	for s.Scan() {
+		name, timestamp, value := s.Metric()
+		// TODO(rartoul): Pool.
+		// Copy name since scanner bytes are recycled.
+		name = append([]byte(nil), name...)
+
+		wg.Add(1)
+		i.opts.WorkerPool.Go(func() {
+			ok := i.write(name, timestamp, value)
+			if ok {
+				i.metrics.success.Inc(1)
+			}
+			wg.Done()
+		})
+
+		i.metrics.malformed.Inc(int64(s.MalformedCount))
+		s.MalformedCount = 0
+	}
+
+	if err := s.Err(); err != nil {
+		logger.Errorf("encountered error during carbon ingestion when scanning connection: %s", err)
+	}
+
+	logger.Debugf("waiting for outstanding carbon ingestion writes to complete")
+	wg.Wait()
+	logger.Debugf("all outstanding writes completed, shutting down carbon ingestion handler")
+
+	// Don't close the connection, that is the server's responsibility.
+}
+
+func (i *ingester) write(name []byte, timestamp time.Time, value float64) bool {
+	datapoints := []ts.Datapoint{{Timestamp: timestamp, Value: value}}
+	// TODO(rartoul): Pool.
+	tags, err := GenerateTagsFromName(name)
+	if err != nil {
+		i.logger.Errorf("err generating tags from carbon name: %s, err: %s",
+			string(name), err)
+		i.metrics.malformed.Inc(1)
+		return false
+	}
+
+	var (
+		ctx     = context.Background()
+		cleanup func()
+	)
+	if i.opts.Timeout > 0 {
+		ctx, cleanup = context.WithTimeout(ctx, i.opts.Timeout)
+	}
+
+	err = i.downsamplerAndWriter.Write(ctx, tags, datapoints, xtime.Second)
+	if err != nil {
+		i.logger.Errorf("err writing carbon metric: %s, err: %s",
+			string(name), err)
+		i.metrics.err.Inc(1)
+		return false
+	}
+
+	if cleanup != nil {
+		cleanup()
```

Comment:
This doesn't get called if `err != nil`.

The common way to do this apparently is:
```go
ctx, cancel = context.WithTimeout(ctx, i.opts.Timeout)

// ...

err = i.downsamplerAndWriter.Write(ctx, tags, datapoints, xtime.Second)
if cancel != nil {
  cancel()
}

// now deal with err
if err != nil {
  // ...
  return
}
```

Reply:
Example here:
https://github.com/uber/tchannel-go/blob/a815a8e7cbcb4ebcd05a73a84f7fd2237dd8a44f/health.go#L127-L131

Reply:
ah thanks, this wasn't broken when I first wrote it but added a hasty refactor. Luckily leaking these the G.C will eventually get them but good catch I'll clean it up proper.